runtime/cudaq/platform/default/rest/CMakeLists.txt

```diff
@@ -9,7 +9,7 @@
 set(INTERFACE_POSITION_INDEPENDENT_CODE ON)
 
 message(STATUS "Building REST QPU.")
-add_library(cudaq-rest-qpu SHARED RemoteRESTQPU.cpp 
+add_library(cudaq-rest-qpu SHARED PyRemoteRESTQPU.cpp 
```

Comment:
Might be that I'm missing some new addition, but as I understand it we do need both the `PyRemoteRESTQPU` and the `RemoteRESTQPU` (with common code extracted out into the `common/BaseRemoteRESTQPU`). The reason for the distinction is to avoid static initializers from LLVM being called twice (Python bindings now link statically to LLVM/MLIR, while `libcudaq-rest-qpu` links to `libcudaq-runtime-mlir` which links statically to LLVM/MLIR) . The `RemoteRESTQPU` remains for our existing runtime `kernel_builder` and AST Bridge workflows, but the new `PyRemoteRESTQPU` is needed for any workflows that target remote REST targets

Reply:
Thanks for the early feedback. Some restructuring and clean up still remains, but, I was planning to have the functionality in single place and remove the 'old/legacy' kernel builder workflow, replace it with the new builder workflow - so for the end users, the interface remains the same (but, the plumbing underneath would be different). Does that seem reasonable? Or am I missing something?

Reply:
The problem I see here is that `libcudaq-rest-qpu` links directly to LLVM/MLIR static libraries. So if you rely on the `RemoteRESTQPU` functionality as part of this library in the Python language bindings, then you will need to link to it (previously we've dynamically loaded it in `LinkedLibraryHolder`). Unfortuntately though, the Python bindings as a library already link statically to LLVM/MLIR, so when this dynamic load happens you will get an error with LLVM static initializers being called more than once. So for any `RemoteRESTQPU` functionality to be used by Python, it has to be decoupled in some way from directly linking to LLVM/MLIR and instead rely on the symbols already loaded in the Python module library (from its static linking to LLVM/MLIR). 

You do need to retain the runtime library's `RemoteRESTQPU` functionality for C++ MLIR-related workflows. The existing solution for this decoupling is the creation of `common/RuntimeMLIRCommonImpl.h` and `common/BaseRemoteRESTQPU.h`. 

I may be wrong and there is some mechanism to enable this unification, but as I see it, we will have to retain `PyRemoteRESTQPU` as separate from the core C++ runtime's `RemoteRESTQPU.cpp`.

Happy to hop on a call to talk in more depth than possible on the chat boxes here